08_project/willshen8/pkg/puppy/store/mapstore.go

```diff
@@ -0,0 +1,73 @@
+package puppy
+
+import (
+	"strconv"
+
+	puppy "github.com/anz-bank/go-course/08_project/willshen8/pkg/puppy"
+)
+
+type MapStore struct {
+	ms     map[uint32]puppy.Puppy
+	nextID uint32
+}
+
+// NewMapStore initialise a new MapStore
+func NewMapStore() *MapStore {
+	return &MapStore{ms: map[uint32]puppy.Puppy{}}
+}
+
+// CreatePuppy create a new puppy and store in mapStore.
+func (m *MapStore) CreatePuppy(p *puppy.Puppy) (uint32, error) {
+	if i, err := strconv.Atoi(p.Value); err == nil {
+		if i < 0 {
+			return 0, &puppy.Error{
+				Message: "Puppy value can't be less than 0.",
+				Code:    puppy.NegativeValue,
+			}
+		}
+	}
+
```

Comment:
you don't handle `err != nil`

Reply:
Sorry, included in lab9 and forgot in lab8. It's included now.

Reply:
sorry, had it lab9, and forgot in lab8

Reply:
I'd write it as 
```go
func (m *MapStore) CreatePuppy(p *puppy.Puppy) (uint32, error) {
	i, err := strconv.Atoi(p.Value)

	if err != nil {
		return 0, &puppy.Error{"Unrecongised puppy value.", puppy.ErrorValueFormat}
	}
	if i < 0 {
		return 0, &puppy.Error{"Puppy value can't be less than 0.", puppy.NegativeValue}
	}

	m.nextID++
	p.ID = m.nextID
	m.ms[p.ID] = *p
	return p.ID, nil
}
```
Rationale being
- Eject as soon as the most obvious error arises (so less chances of nesting creep)
- Avoid `if` nesting like the plague (so it's easy to read)
- No struct field names (makes things one-line and easier to read)

Reply:
Cheers, that's exactly how the code turned out to be in the later [labs](https://github.com/anz-bank/go-course/blob/master/10_rest/willshen8/pkg/puppy/mapstore.go).